tokio-postgres/src/copy_both.rs

```diff
@@ -0,0 +1,382 @@
+use crate::client::InnerClient;
+use crate::codec::FrontendMessage;
+use crate::connection::RequestMessages;
+use crate::{simple_query, Error};
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+use futures::channel::mpsc;
+use futures::stream::FusedStream;
+use futures::{ready, Sink, SinkExt, Stream, StreamExt};
+use log::debug;
+use pin_project_lite::pin_project;
+use postgres_protocol::message::backend::Message;
+use postgres_protocol::message::frontend;
+use postgres_protocol::message::frontend::CopyData;
+use std::marker::{PhantomData, PhantomPinned};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// The state machine of CopyBothReceiver
+///
+/// ```ignore
+///             CopyBoth
+///             /   |   \
+///        ,---'    |    `---,
+///       /         |         \
+///      v          v          v
+///  CopyOut-->ServerError<--CopyIn
+///      \          |          /
+///       `---,     |     ,---'
+///            \    |    /
+///             v   v   v
+///              CopyNone
+///                 |
+///                 v
+///            CopyComplete
+///                 |
+///                 v
+///           CommandComplete
```

Comment:
I don't think this is the right place in the state machine for `ServerError`. If a server error happens in `CopyBoth` mode, that's the odd (and hard-to-test) case where an error is thrown by the server in the middle of streaming. From the docs, it looks like Postgres will just return the `ErrorResponse` and then `ReadyForQuery`, without either of the `CommandComplete` messages.

You can test this by hacking up `test_decoding` to throw an error randomly every 10 records or something.